File: Two Sum/1 Two Sum 1-7-24.py
""" 
    In this implementation of a solution to the Two Sum problem we iterate over an array while
    looking for possible combinations of the numbers in the array whose sum equate to the target.
    We can do this by taking the value of each value, finding the number that complements the
    value to equate to the target, and then looking for the compliment in the array. This question
    will only input arrays that have one solution so we can stop as soon as we find the answer.
    
    To solve we iterate over the array, find the complement, and search the array for this compliment
    With this implementation, however, we will be cutting the search down time by looking through a
    dictionary that we create as we search for the solution. This means that the worst case scenario
    will be the same as if we search the whole array each time, but if the target and complement are
    found earlier in the list we save time and space.
"""
def two_sum(nums, target):
    """ Create dictionary that we will add to while searching """
    num_dict = {}

    """ Iterate over the aray and find the complement of the interation """
    for i, num in enumerate(nums):
        complement = target - num

        """ Check the dictionary for the complement """
        if complement in num_dict:
            """ If the dicitonary contains the complement then we are done """
            """ Return the indices of the complement and the original number """
            return [num_dict[complement], i]

        """ If the complement doesn't exist then add the value and it's index to the dictionary """
        num_dict[num] = i

    """ If not found, return empty string """
    return []

""" Test """
nums = [2, 7, 11, 15]
target = 9
result = two_sum(nums, target)
print(result)

File: Problems/Range of Sum BST/Two Sum/1TwoSum1-7-24.js
/* Two Sum */
/* When given an array of integers and a target integer, return the two numbers */
/* indices that add to result in the target integer                             */

/* Nums = Array of integers  */
/* Target = Target integer for the sum */
function twoSum(nums, target) {
    
    /* Use a map (key:value pair array) in this implimentation of the solution in order to find target */
    /* Using a map allows us to store the array in which the value of the keys are the indices we are  */
    /* Looking for */
    const numMap = new Map();

    /* Now, we look through the array of numbers to find two the two numbers that add to be the sum.
       while we look, we have to acknowledge the number of the iteration that we are on as well as 
       the number that complements the number of the iteration. We do this by implementing a 
       map. This map will house the array as the value of the array and it's position. We will iterate
       over the array as well as make a map until we have found a value that whose complement is in the
       map that we create. */
    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];

        
        if (numMap.has(complement)) {
            /* If the map that we have created includes the complement number then we are done, we have both values
                 */
            return [numMap.get(complement), i];
        }

        /* If the complement hasn't been found, add the number and it's index to the map and iterate again */
        numMap.set(nums[i], i);
    }

    /* If there is no two numbers that can return the sum simply return an empty array */
    return [];
}

/* Test */
const nums = [2, 7, 11, 15];
const target = 9;
const result = twoSum(nums, target);
console.log(result); /* The result output will be a small array with the two indices of the numbers whose sum equate to the target */